back/src/Model/GameRoom.ts

```diff
@@ -249,32 +246,80 @@ export class GameRoom {
+            let hasKickOutSomeone = false;
+
+            const previewNewGroupPosition = previewGroupPosition(headMembers);
+
+            if (!previewNewGroupPosition) {
+                this.leaveGroup(user);
+                return;
+            }
+
+            if (user.following) {
+                const followingMembers = group
+                    .getUsers()
+                    .filter((currentUser) => currentUser.following === user.following);
+
+                if (group.getUsers().length - 1 === followingMembers.length) {
+                    let isOutOfBounds = false;
+                    for (const member of followingMembers) {
+                        const distance = GameRoom.computeDistanceBetweenPositions(
+                            user.getPosition(),
+                            previewNewGroupPosition
+                        );
+
+                        if (distance > this.groupRadius) {
+                            isOutOfBounds = true;
+                            break;
+                        }
+                    }
+                    group.setOutOfBounds(isOutOfBounds);
                 }
-                const usrPos = foreignUser.getPosition();
-                const grpPos = foreignUser.group.getPosition();
-                const distance = GameRoom.computeDistanceBetweenPositions(usrPos, grpPos);
-
-                if (distance > this.groupRadius) {
-                    if (foreignUser.hasFollowers() || foreignUser.following) {
-                        // If one user is out of the group bounds BUT following, the group still exists... but should be hidden.
-                        // We put it in 'outOfBounds' mode
-                        group.setOutOfBounds(true);
-                        noOneOutOfBounds = false;
-                    } else {
-                        this.leaveGroup(foreignUser);
+            } else if (user.hasFollowers()) {
+                const followingMembers = group.getUsers().filter((currentUser) => currentUser.following === user);
+                let isOutOfBounds = false;
+
+                if (followingMembers.length === group.getUsers().length - 1) {
+                    for (const member of followingMembers) {
+                        const distance = GameRoom.computeDistanceBetweenPositions(
+                            member.getPosition(),
+                            previewNewGroupPosition
+                        );
+
+                        if (distance > this.groupRadius) {
+                            isOutOfBounds = true;
+                            break;
+                        }
+                    }
+
+                    group.setOutOfBounds(isOutOfBounds);
+                }
+            } else {
+                for (const headMember of headMembers) {
+                    if (!headMember.group) {
+                        this.leaveGroup(headMember);
+                        headMembers = group.getGroupHeads();
+                        return;
```

Comment:
The return looks weird here. Especially after assigning the headMembers variable that is not used.
I don't get what you are trying to do? Maybe adding comments would help?